add CSS partial tagged template literal

nicholasrice opened the pull request:
<!---
Thanks for filing a pull request 😄 ! Before you submit, please read the following:

Search open/closed issues before submitting. Someone may have pushed the same thing before!

Provide a summary of your changes in the title field above.
-->

# Pull Request

## 📖 Description
This PR adds a new CSS tagged template literal called `cssPartial`. `cssPartial` allows abstracting CSS blocks that are not valid CSS in-and-of themselves into a structure usable by `ElementStyles`. `cssPartial` allows interpolation of `CSSDirective` instances as well as strings, and itself returns a `CSSDirective` that can be interpolated into `ElementStyles`.

closes #4633
<!---
Provide some background and a description of your work.
What problem does this change solve?
Is this a breaking change, chore, fix, feature, etc?
-->

### 🎫 Issues

<!---
* List and link relevant issues here.
-->

## 👩‍💻 Reviewer Notes

<!---
Provide some notes for reviewers to help them provide targeted feedback and testing.

Do you recommend a smoke test for this PR? What steps should be followed?
Are there particular areas of the code the reviewer should focus on?
-->

## 📑 Test Plan

<!---
Please provide a summary of the tests affected by this work and any unique strategies employed in testing the features/fixes.
-->

## ✅ Checklist

### General

<!--- Review the list and put an x in the boxes that apply. -->

- [ ] I have included a change request file using `$ yarn change`
- [x] I have added tests for my changes.
- [x] I have tested my changes.
- [ ] I have updated the project documentation to reflect my changes.
- [x] I have read the [CONTRIBUTING](https://github.com/Microsoft/fast/blob/master/CONTRIBUTING.md) documentation and followed the [standards](https://www.fast.design/docs/community/code-of-conduct/#our-standards) for this project.

### Component-specific

<!--- Review the list and put an x in the boxes that apply. -->
<!--- Remove this section if not applicable. -->

- [ ] I have added a new component
- [ ] I have modified an existing component
- [ ] I have updated the [definition file](https://github.com/Microsoft/fast/blob/master/packages/web-components/fast-components/CONTRIBUTING.md#definition)
- [ ] I have updated the [configuration file](https://github.com/Microsoft/fast/blob/master/packages/web-components/fast-components/CONTRIBUTING.md#configuration)

## ⏭ Next Steps

<!---
If there is relevant follow-up work to this PR, please list any existing issues or provide brief descriptions of what you would like to do next.
-->

EisenbergEffect commented on packages/web-components/fast-element/docs/guide/leveraging-css.md:
```diff
@@ -128,6 +128,71 @@ Rather than simply concatenating CSS strings, the `css` helper understands that
 :::
 
+### Partial CSS
+There are times when you may want to create reusable blocks of *partial* CSS, where the abstraction is not valid CSS in and of itself, such a groups of CSS properties or a complex value. To do that, you can use the `cssPartial` tagged template literal: 
```
```suggestion
There are times when you may want to create reusable blocks of *partial* CSS, where the abstraction is not valid CSS in and of itself, such as groups of CSS properties or a complex value. To do that, you can use the `cssPartial` tagged template literal: 
```

chrisdholt replied:
nice - my mind walked right past this!

EisenbergEffect commented:
@nicholasrice One quick thought/question...is there any way to factor out and share code between the `css` and `cssPartial` helpers? If we add bindings in CSS, I'm thinking we'd need to add it in both places as well, so I'm wondering if there's some way to share the basic parsing/collecting code.

nicholasrice commented:
We can probably share some, though `css` handles a few extra cases which makes re-use a bit tricky.

nicholasrice commented:
Hmm I just realized `CSSDirective` returns `ComposableStyles` and not just string, so this won't work as-is. Converting to draft

nicholasrice commented:
@EisenbergEffect Okay I updated this to support all valid return types from `CSSDirective` and also to re-use style collection logic from `css`. 

The main difference between `css` and `cssPartial` is that `css` creates a new CSSStyleSheet for every string in the reduced styles, whereas `cssPartial` will zip-up all strings into a single string and return that from the resulting `CSSDirective`. This means it supports some rather strange syntaxes such as:

**This sets a host's color to blue:**
```ts
class MyDirective extends CSSDirective {
    createCSS() {
        return css`:host { color: blue; }`
    }
}
const partial = cssPartial`color: ${new MyDirective} green`;

css`
    :host {
        ${partial}
    }
`
```

A few questions:
1. Are there general concerns with enabling the above scenario?
2. I currently have `cssPartial` restricted to embedding `string | CSSDirective` (no `ElementStyles or CSSStyleSheet`), but this can obviously be circumvented by returning an `ElementStyles` or `CSSStyleSheet` from a `CSSDirective`. Should I just open up the API?
3. If we changed the behavior of `css` to work like `cssPartial` I think we could technically support both scenarios with one template function, though I think we would loose some of the `CSSStyleSheet` re-use that `ElementStyles` is optimizing for internally. We probably *don't* want to do this, but opening the conversation just in case having one function that does both is preferred.

EisenbergEffect commented:
Can you expand a bit more on your third point above? I think it would be nice to have a single function, but not sure I fully understand what we'd give up.

nicholasrice commented:
https://github.com/microsoft/fast/blob/master/packages/web-components/fast-element/src/styles/element-styles.ts#L150 is where ElementStyles converts all style *strings* into CSSStyleSheet instances, caching the instances. That logic is why we can't provide incomplete or invalid CSS to `css` (such as ``` css`color: red` ```). When we try to populate a `CSSStyleSheet` with `"color: red"`, the browser rightfully doesn't know how to convert that to a CSS rule because there is no selector, so it just does nothing.

 If, instead, `ElementStyles` zipped all of the strings into *one* string (or at least all immediate siblings, for instance `[string, string, CSSStyleSheet, string, CSSStyleSheet] => [string, CSSStyleSheet, string, CSSStyleSheet]`), then including a incomplete rule like the above into another `ElementStyles` would just cause the strings to get zipped together prior to `CSSStyleSheet` hydration.

I *think* doing so would reduce the amount of CSSStyleSheet re-use we could do though, because we can no-longer always leverage `CSSStyleSheets` for the individual style strings in a `CSSStyleSheet`.

EisenbergEffect commented:
Ah, yes. That makes sense. It seems like it could be a perf cost since the browser won't be sharing the sheets or be able to dedupe them internally. So, probably we don't want to do that.

Hmm. I'm not coming up with anything better than the `cssPartial` template function at the moment. If ElementStyles or a CSSStyleSheet are returned, will that break things?

nicholasrice commented:
Returned from where? `cssPartial`?

EisenbergEffect commented:
Yes. You mentioned expanding the types to allow that I think. If those were allowed, would it break things?

nicholasrice commented:
No, it would work the same. It would just make it a little easier to write strange CSS like the above.

EisenbergEffect commented:
Ok. I think it's ok then. Let's just not put the strange stuff in the docs 😉

nicholasrice commented:
heh yea okay. I'll update the function.

janechu commented on packages/web-components/fast-element/docs/guide/leveraging-css.md:
```diff
@@ -128,6 +128,71 @@ Rather than simply concatenating CSS strings, the `css` helper understands that
+import { css, cssPartial } from "@microsoft/fast-element";
+
+const partial = cssPartial`color: red;`;
+const styles = css`:host{ ${cssPartial} }`;
```
```suggestion
const styles = css`:host{ ${partial} }`;
```

janechu commented on packages/web-components/fast-element/docs/guide/leveraging-css.md:
````diff
@@ -128,6 +128,71 @@ Rather than simply concatenating CSS strings, the `css` helper understands that
+The `CSSDirective` can then be used in an `ElementStyles`, where the CSS string from `createCSS()` will be interpolated into the stylesheet, and the behavior returned from `createBehavior()` will get bound to the element using the stylesheet:
+
+```ts
+const styles = css`:host {${new RandomWidth()}}`;
````
Is the intended result that `CSSDirective` class will execute the `crreateCSS()` in the constructor?

nicholasrice replied:
`createCSS()` gets called by `ElementStyles`, so that happens implicitly

janechu reviewed (approved):
Couple clarifying questy
